### PR TITLE
Fix/memstreamriver

### PR DESCRIPTION
**Pull request by NicolasNoya**
## Add MemStream: Memory-Based Streaming Anomaly Detection

This PR introduces **MemStream**, a state-of-the-art online anomaly detection framework designed for high-dimensional data streams with concept drift, based on the paper ["MemStream: Memory-Based Streaming Anomaly Detection"](https://arxiv.org/pdf/2106.03837) by Bhatia et al.

#### Core Implementation
- **`MemStream` (Base Class)**: Abstract base class providing the core framework for memory-based anomaly detection
- **`MemStreamPCA`**: Concrete implementation using PCA-based feature encoding

### Architecture

The implementation consists of two main components:

1. **Feature Encoder**: Transforms high-dimensional inputs into lower-dimensional representations
   - Currently implements PCA-based projection
   - Extensible design allows for future encoders (denoising autoencoders and information bottleneck, not implemented due to compatibility issues)

2. **Memory Module**: Maintains a dynamic collection of encoded "normal" data representations
   - Adapts to concept drift without explicit labels
   - Configurable replacement strategies: FIFO, LRU, and Random
   - Prevents memory poisoning from anomalous samples

### Key Features

- **Online Learning**: Processes data points one at a time
- **Unsupervised Detection**: No labels required during inference (optional during training)
- **Concept Drift Adaptation**: Memory evolves over time to handle distribution changes
- **Flexible Scoring**: Uses k-nearest neighbors with exponential weighting to compute anomaly scores
- **Grace Period**: Collects initial samples to bootstrap the encoder before scoring begins
- **Memory Management**: Configurable size and replacement policies

### Parameters

- `memory_size`: Maximum number of encoded normal samples to store (default: 1,000 for PCA variant)
- `max_threshold`: Threshold for accepting samples into memory (default: 0.1)
- `grace_period`: Number of initial samples before scoring begins (default: 5,000)
- `n_components`: Number of PCA components (default: 20) (coded to take the value that makes PCA possible if n_components is inappropriate)
- `k`: Number of nearest neighbors for scoring (default: 5)
- `gamma`: Exponential weighting factor (default: 0.1)
- `replace_strategy`: Memory replacement policy (FIFO, LRU, or RANDOM)

### CODE REVIEW

**Comment by kulbachcedric** [comment]
Hi @NicolasNoya,

that's a nice contribution! 
However, I think the setup.py creates some issues here. 
Can we remove it? 

Best
Cedric

File: river/anomaly/memstream.py [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+    IB = "information_bottleneck"  # TODO: implement
+
+
+class ReplaceStrategy:
+    FIFO = "FIFO"  # First In, First Out
+    LRU = "LRU"  # Least Recently Used
+    RANDOM = "RANDOM"  # Random replacement
+
+
+class MemStream(anomaly.base.AnomalyDetector):
```
Comment (kulbachcedric) [inline river/anomaly/memstream.py]: I would remove this class and integrate it into the MemStreamPCA class, as we currently have only on Memstream variant implemented.

**Comment by kulbachcedric** [comment]
Hi @NicolasNoya 
nice one! 
Actually the changes within the setup.py still appear within the changes. 
I would suggest add add docstrings to every function you are adding.

For the Benchmarks, we could create a new issue?

Best
Cedric

File: setup.py
[the remark this answers is not in the document]
Comment (kulbachcedric) [reply setup.py]: @NicolasNoya

**Comment by kulbachcedric** [comment]
Hi @NicolasNoya,
just wanted to reach out to you again. 
Would love to have your contribution in river. 
With some minor changes, I think we can merge it ;-) 

Best
Cedric

**Comment by NicolasNoya** [comment]
Hey @kulbachcedric!! Sorry for not answering in a while, my last month in the University was super intense and I couldn't work on the code (then I took some vacations). I will try to fix the code this weekend and maybe I can start with other things that the library requires, I'll have a little more free time now that I've finish my studies.

Thanks for reaching out!
Nicolas Noya

**Comment by NicolasNoya** [comment]
Hello @kulbachcedric, I did the modifications that you asked! Please let me know if you’d like any further modifications or additional testing. :smile:

**Comment by kulbachcedric** [comment]
> Hello @kulbachcedric, I did the modifications that you asked! Please let me know if you’d like any further modifications or additional testing. 😄

Hi @NicolasNoya,
thanks! It looks good for me. On minor issue is the failing ruff test. 
Could you please run it (`make format`) and commit the changes ruff proposes? 

Best
Cedric

**Comment by kulbachcedric** [comment]
Hi @NicolasNoya,
are there any news on this one `uv run make format` would do the trick already.

Best
Cedric

**Comment by MaxHalford** [comment]
Sorry but our policy is against using batch methods like sklearn's PCA in the hot loop. We can reconsider this method once we've implemented streaming PCA.
